### Single read PCA manual
# install.packages('vcfR')
# install.packages('SNPfiltR', lib = "/global/scratch/users/arphillips/toolz/R/")
library(vcfR)
library(stringr)
library(SNPfiltR, lib.loc = "/global/scratch/users/arphillips/toolz/R/")

# Load VCF
vcf <- read.vcfR("/global/scratch/users/arphillips/spectral_aspen/data/processed/filtered_snps/rad_aspen.all.depth.3dp30.nocall.vcf.gz", verbose = FALSE )

str(vcf)
vcf@gt[1:5,1:5]

### Any additional SNP filtering
## remove invariant sites generated by dropping genotypes
dp <- extract.gt(vcf, element = "DP", as.numeric = T)
hist(dp)

vcf <- min_mac(vcf, min.mac = 1)

## linkage?
#linkage filter vcf to thin SNPs to one per 500bp
# vcfR.thin<-distance_thin(vcf, min.distance = 500)

# create vector of counts of ref and alt read count
# queryMETA(vcf, element = "AD")
ad <- extract.gt(vcf, element = "AD", as.numeric = F)
ad[1:5,1:5]
str(ad)

ref <- masplit(ad, record = 1, sort = 0)
ref[1:5,1:5]
dim(ref)
alt <- masplit(ad, record = 2, sort = 0)
alt[1:5,1:5]
dim(alt)

# draw read and create matrix
genotypes = dim(ref)[2]
snps = dim(ref)[1]

single <- matrix(NA, nrow = snps, ncol = genotypes)

for (i in 1:genotypes){
  for (j in 1:snps){
    ref_count = ref[j,i]
    alt_count = alt[j,i]
    if (ref_count + alt_count == 0){
      single[j,i] <- NA
    }
    else{
      reads <- c( rep(0, ref_count), rep(1, alt_count) )
      set.seed(7) # set.seed before drawing
      single[j,i] <- sample(reads, 1) 
    }
  }
}

single[1:5,1:5]
sum(is.na(single))
dim(single)

## Test code ##
# reads <- c( rep(0, ref[4,1]), rep(1, alt[4,1]) )
# set.seed(7) # set.seed before drawing
# sample(reads, 1)

# calculate allele frequencies based on single reads 
# f = (count minor allele reads)/(count of major + minor allele reads)

# calculate covariance matrix M
# See ANGSD documentation

# calculate eigen vectors of M
#eigen(M)

# Plot PCA results